compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3700,8 +3703,143 @@
         }
         BFutureType futureType = new BFutureType(TypeTags.FUTURE, type, null);
         checkExpr(expr, env, futureType);
+        setEventualTypeForExpression(expr, type);
     }
 
+    // eventual type if not directly referring a worker is T|error. future<T> --> T|error
+    private void setEventualTypeForExpression(BLangExpression expression,
+                                              BType currentExpectedType) {
+        if ((expression == null) || (expression.getKind() != NodeKind.SIMPLE_VARIABLE_REF)) {
```

Comment:
I don't think we can limit this for simple_variable_references, future could be coming from a map, array, from a invocation etc. 
```ballerina
function bar() returns future<int> {
    function () returns int f = () => 0;
    return start f();
}

public function main() {
    int x = wait bar();
}
```

Reply:
Oh my bad, I was looking at wrong function. Please disregard.